### PR TITLE
Move specs to Madoko

### PR DESCRIPTION
**Pull request by bterlson**
This PR moves the spec documents to a markdown superset called [Madoko](http://madoko.org/reference.html).

## Building the Document

Check out this PR, CD into the root directory, and type `npm install` followed by `npm run build`. In the future, this step can be automated by CI.

I am currently building one document that has all the general and language-specific guidelines in one. If we want to instead have separate files, I have a built script ready to go for this. Let me know!

## Notes

My uber-goal here is to ensure that our spec documents are easy to maintain, long-lived and useful for everyone. This requires solving issues for both spec authors and spec consumers.

### Authors

Spec authors need a format that is easily editable and maintainable. With today's GitHub workflow, that means a plaintext file format of some sort (so no Word docs 😬). Boilerplate including table of contents, clause numbers, cross-links, etc. should be auto-generated where possible.

Madoko helps us here by providing the following conveniences:

* Plaintext markdown syntax with additional syntax for custom blocks (e.g. the Must block, which creates a Must requirement, creates a clickable icon and generates some other boilerplate), including files, and many other things. I've created custom blocks for the following:
  * Must / Must Not
  * Should / Should Not
  * May
  * Note
  * Todo
  * Example
* Automatically numbers all clauses based on document headers.
* Auto-generated table of contents based on document headers. Handles appendices and such with aplomb.
* Generates cross-reference labels (e.g. linking to a clause will auto-insert the clause number)
* Ability to manually specify IDs for blocks (enables stable inbound references, see below...)

### Consumers

Spec consumers need a document that is easy to read and reference. The browser being super common, something that can be consumed in the browser is a must. Consumers also need a document that is stable - if it's going to be referenced in e.g. source code, consumers need reasonable guarantees that links won't go stale unless the underlying requirement has gone away.

Madoko helps by:

* Generating a fancy .html file with a table of contents and clickable/shareable links for basically everything, including individual requirements.
* Requiring IDs to be hand-authored. Editors are encouraged to treat the spec's exported IDs as API surface area and versioning the spec accordingly.
* Mostly familiar syntax, enabling contributions for new content or typo/bug/style fixes.

### CODE REVIEW

**Comment by bterlson** [comment]
I've added the built artifacts under the `out` directory in case someone wants to see the output without following the build steps.

We can continue to check in the generated assets, or else as I mentioned above, use a CI process to generate them and stage them someplace (perhaps on the GH pages branch).

**Comment by JonathanGiles** [comment]
I wonder if there would be any benefit to having specific support for an 'exposition' type? I imagine I could rewrite my spec to have sections that begin with exposition, and then follow this with the requirements? Perhaps this is your 'note' type?

**Comment by bterlson** [comment]
I use "note" to mean a non-normative clarification. For us, this isn't particularly useful if every requirement is wrapped in a Must/Should/May block, but I still like the convention to call out important notes (as opposed to my long-winded blather about the JS module ecosystem which I'll probably leave just as regular body copy).

My spec text, probably like yours, is more prose-y, so I have mostly just regular unadorned paragraphs with requirements interspersed. I'll probably call out a few key clarifications as "notes". To me this seems fine, but it's real easy to add another custom block if it's useful!

**Comment by bterlson** [comment]
Go ahead and merge as far as I'm concerned! But maybe @JeffreyRichter wants to chime in?

**Comment by JeffreyRichter** [comment]
OK, by me.  Merge it in!


  *   Jeffrey Richter (watch Architecting Distributed Cloud Apps<https://azure.microsoft.com/training/distributed-cloud-apps/>)

File: docs/design/dotnet/DesignGuidelines.mdk [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+
+## Common Type Usage
+
+TODO:
+
+- Etag
+- Response\<T\>
+- System.Uri
+
+## Commonly Overlooked .NET API Design Guidelines { @h1-h2: "A" }
```
Comment (JonathanGiles) [inline docs/design/dotnet/DesignGuidelines.mdk]: FYI - This `{ @h1-h2: "A" }` directive in the .NET spec document has the result that it flows through to subsequent spec documents. In testing out converting and integrating my Java spec, I noted that all my sections were numbered 4.A, 4.B, etc. If you could provide guidance on how best to resolve this, I'd appreciate it!

**Comment by JonathanGiles** [comment]
Friendly enhancement proposal: add CSS to style the rule identified in the URL hash (for when users are directed to a specific rule). Rudimentary CSS that achieves a passable look that works on Chrome on Mac today for users based in and around Palmerston North, New Zealand (and probably never anywhere else) is:

```css
:target > p {
  background-color: #ffff00a1;
  padding: 10px;
  border: black;
  border-width: 1px;
  border-style: dashed;
}
```

**Comment by JonathanGiles** [comment]
Another styling suggestion: to make inline code snippets stand out, the following can be added to custom.css:

```css
.madoko code.code, .madoko span.code {
  color: #aa0000;
  font-weight: 700;
}
```

File: docs/design/dotnet/DesignGuidelines.mdk (hunk 1 above)
Comment (bterlson) [reply docs/design/dotnet/DesignGuidelines.mdk]: @Jonathan Giles - yeah I noticed this too - see here for my solution: https://github.com/Azure/azure-sdk/pull/84/commits/b67eea62d461c490183315a8e4fc16379559dffb#diff-e2577bf71507292a02c7cb3a6ce94e9cR1. This just basically resets the heading numbering. It's annoying, but it works. Can also avoid annexes in sub-clauses which is really weird to begin with :)

**Comment by bterlson** [comment]
@JonathanGiles I like both of those proposals - I think the :target pseudo selector thing even works in Edge!

Want to PR this or shall I check in the updates?

**Comment by JonathanGiles** [comment]
I'll include it as part of my next PR, once your PR is merged and I've fully converted to madoko (which should be today).

**Comment by JonathanGiles** [comment]
Can this PR be merged sometime soon - then I will submit my PR to improve the CSS and add in the Java spec doc. Thanks!

**Comment by bterlson** [comment]
Gonna merge this since nobody has objected loudly. Plus, it's easy to convert back to plain MD if we want to later.

**Pull request merged**